pandas/core/groupby/ops.py

```diff
@@ -1243,7 +1244,7 @@ def groupings(self) -> list[grouper.Grouping]:
         ping = grouper.Grouping(lev, lev, in_axis=False, level=None)
         return [ping]
 
-    def _aggregate_series_fast(self, obj: Series, func: Callable) -> np.ndarray:
+    def _aggregate_series_fast(self, obj: Series, func: Callable) -> NoReturn:
```

Comment:
When we have `NotImplementedError`, does the `NoReturn` not create issues with subtypes? (in general, no checked this particular case)

Reply:
Thanks. yes it does appear that the exception is not for public consumption.

Reply:
NoReturn is useful in cases like this to narrow the type:

```py
def error() -> NoReturn: ...

def test(x: int | str) -> str:
    # x: int | str
    if isinstance(x, int):
        error()
    # x: str
    return x + "some_string"
```

I don't there are many places where NoReturn can be used in pandas, in most cases we should make classes abstract.

Is there a historic reason to avoid inheriting from `ABC` and using `abstractmethod`? Happy to make PRs to make more methods abstract.